unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+                Some(m)
+            }
+        }
+    }
+
+    /// Returns the total size of the file this entry represents, or none if not a file.
+    pub fn total_file_size(&self) -> Option<u64> {
+        use Entry::*;
+        match self {
+            File(_, _, _, sz) => Some(*sz),
+            _ => None,
+        }
+    }
+
+    /// Returns the current Cid unelss for root elements.
+    pub fn cid(&self) -> Option<&Cid> {
+        use Entry::*;
+        match self {
+            RootDirectory(_, _) | File(None, _, _, _) | Symlink(None, _, _) => None,
+            Bucket(cid, _)
+            | Directory(cid, _, _)
+            | File(Some(cid), _, _, _)
+            | Symlink(Some(cid), _, _) => Some(cid),
+        }
+    }
+}
+
+impl InnerEntry {
+    fn new_root_dir(_: Cid, metadata: Metadata, name: &str, depth: usize) -> Self {
+        let mut path = PathBuf::new();
+        path.push(name);
+        Self {
+            kind: InnerKind::RootDirectory,
+            path,
+            metadata,
+            depth,
+        }
+    }
+
+    fn new_root_bucket(_: Cid, metadata: Metadata, name: &str, depth: usize) -> Self {
+        let mut path = PathBuf::new();
+        path.push(name);
+        Self {
+            kind: InnerKind::BucketAtRoot,
+            path,
+            metadata,
+            depth,
+        }
+    }
+
+    fn new_root_file(
+        _: Cid,
+        metadata: Metadata,
+        name: &str,
+        step: Option<FileVisit>,
+        file_size: u64,
+        depth: usize,
+    ) -> Self {
+        let mut path = PathBuf::new();
+        path.push(name);
+        Self {
+            kind: InnerKind::File(None, step, file_size),
+            path,
+            metadata,
+            depth,
+        }
+    }
+
+    fn new_root_symlink(_: Cid, metadata: Metadata, name: &str, depth: usize) -> Self {
+        let mut path = PathBuf::new();
+        path.push(name);
+        Self {
+            kind: InnerKind::Symlink(None),
+            path,
+            metadata,
+            depth,
+        }
+    }
+
+    pub fn as_entry(&self) -> Entry<'_> {
+        use InnerKind::*;
+        match &self.kind {
+            RootDirectory | BucketAtRoot => Entry::RootDirectory(&self.path, &self.metadata),
+            RootBucket(cid) => Entry::Directory(cid, &self.path, &self.metadata),
+            Bucket(cid) => Entry::Bucket(cid, &self.path),
+            Directory(cid) => Entry::Directory(cid, &self.path, &self.metadata),
+            File(cid, _, sz) => Entry::File(cid.as_ref(), &self.path, &self.metadata, *sz),
+            Symlink(cid) => Entry::Symlink(cid.as_ref(), &self.path, &self.metadata),
+        }
+    }
+
+    fn set_path(&mut self, name: &str, depth: usize) {
+        while self.depth >= depth && self.depth > 0 {
+            assert!(self.path.pop());
+            self.depth -= 1;
+        }
+
+        self.path.push(name);
+        self.depth = depth;
```

Comment:
I seem to have forgotten the `assert_eq!(self.depth, self.path.ancestors.count());` pre and post conditions off while losing my mind over the different cases. These work now, so adding them back in as debug_asser_eq.

Reply:
Done